engine/src/witness/hub.rs

```diff
@@ -0,0 +1,367 @@
+mod hub_chain_tracking;
+mod hub_deposits;
+mod hub_source;
+
+use cf_chains::dot::{
+	PolkadotAccountId, PolkadotBalance, PolkadotExtrinsicIndex, PolkadotHash, PolkadotSignature,
+	PolkadotTransactionId, PolkadotUncheckedExtrinsic,
+};
+use cf_primitives::{EpochIndex, PolkadotBlockNumber};
+use futures_core::Future;
+use state_chain_runtime::AssethubInstance;
+use subxt::{
+	backend::legacy::rpc_methods::Bytes,
+	config::PolkadotConfig,
+	events::{EventDetails, Phase, StaticEvent},
+	utils::AccountId32,
+};
+
+use tracing::error;
+
+use std::{collections::BTreeSet, sync::Arc};
+
+use cf_utilities::task_scope::Scope;
+
+use crate::{
+	db::PersistentKeyDB,
+	dot::retry_rpc::{DotRetryRpcApi, DotRetryRpcClient},
+	state_chain_observer::client::{
+		extrinsic_api::signed::SignedExtrinsicApi,
+		storage_api::StorageApi,
+		stream_api::{StreamApi, FINALIZED},
+		STATE_CHAIN_CONNECTION,
+	},
+	witness::common::chain_source::extension::ChainSourceExt,
+};
+use anyhow::Result;
+pub use hub_source::{HubFinalisedSource, HubUnfinalisedSource};
+
+use super::common::{
+	chain_source::Header,
+	epoch_source::{EpochSourceBuilder, Vault},
+};
+
+// To generate the metadata file, use the subxt-cli tool (`cargo install subxt-cli`):
+// subxt metadata --pallets Proxy,Balances,TransactionPayment,System --url
+// wss://asset-hub-polkadot-rpc.dwellir.com:443 > metadata.assethub.scale
+#[subxt::subxt(runtime_metadata_path = "metadata.assethub.scale")]
+pub mod assethub {}
+
+#[derive(Debug, Clone)]
+pub enum EventWrapper {
+	ProxyAdded { delegator: AccountId32, delegatee: AccountId32 },
+	Transfer { to: AccountId32, from: AccountId32, amount: PolkadotBalance },
+	TransactionFeePaid { actual_fee: PolkadotBalance, tip: PolkadotBalance },
+	ExtrinsicSuccess,
+}
+
+use assethub::{
+	balances::events::Transfer, proxy::events::ProxyAdded, system::events::ExtrinsicSuccess,
+	transaction_payment::events::TransactionFeePaid,
+};
+
+pub fn filter_map_events(
+	res_event_details: Result<EventDetails<PolkadotConfig>, subxt::ext::subxt_core::error::Error>,
```

Comment:
?
```suggestion
	res_event_details: Result<EventDetails<AssethubConfig>, subxt::ext::subxt_core::error::Error>,
```

Reply:
we might want to define an explicit Config for Assethub so it's not tied to the Polkadot one defined in subxt

Reply:
In `witness/hub.rs` we only use the Polkadot Config in order to pass it to the `EventDetails<>` and `Events<>` structs, which themselves access the associated `PolkadotConfig::Hash` type or just put the Config in a phantom data, respectively.

Its more interesting in the (dot) `rpc` file(s) where we also access the associated `PolkadotConfig::Header` type.

The `PolkadotConfig` struct itself inherits all fields from `SubstrateConfig` (https://github.com/paritytech/subxt/blob/f5e9ce0d2ca59ecd716027f3faac7fcc0875ac1b/core/src/config/polkadot.rs#L19-L28). 

I think there are two options how to proceed:
1. Make the rpc functions parametric over the config type, i.e. `function<T: Config>(...)`, and instantiate it with `PolkadotConfig` for dot and with a newly created `AssethubConfig` for assethub.
2. Rename all occurences of `PolkadotConfig` in the rpc files to `SubstrateConfig` and create a type alias `type AssethubConfig = SubstrateConfig`.

Reply:
As discussed, think we can leave it as is, or use option 2. If we need to update we'll need to update it whether we parameterise or not, so doesn't make much of a difference in the end.